x-pack/plugins/security_solution/docs/prebuilt_rules_customization_rfc.md

```diff
@@ -0,0 +1,2857 @@
+        <th style="border-right:3px solid black">Target version</th>
+        <th>Merged version (output)</th>
+      </tr>
+    </thead>
+    <tbody align="center">
+      <tr>
+        <td style="border-right:3px solid black">Keep customizations if conflict</td>
+        <td><code>50</code></td>
+        <td><code>65</code></td>
+        <td style="border-right:3px solid black"><code>20</code></td>
+        <td><code>65</code></td>
+      </tr>
+    </tbody>
+  </table>
+
+##### Array fields
+
+###### Array of strings fields
+
+For **array of strings** fields, conflict resolution will take place on an element by element basis - that means that the algorithm will not consider or try to solve changes within a single element. For example:
+
+```
+base: [test1]
+current: [my-test1]
+target: [test2]
+output: [my-test1]
+```
+Therefor the example above, the algorithm should not try to merge the single string element by breaking it into parts.
+
+Example **array of strings** fields: `index`, `tags`, `references`,
+
+###### Array of objects fields
+
+Since the properties in each object within the array are highly dependent on one another, the whole object should be treated as a block, and not attempt to diff or merge changes to individual fields within each element.
+
+For example: in a `required_fields` update, if the `type` is updated from `unknown` in the base version to `keyword` in the `target` version, but in the same object the user has updated the base `name` of the field, the change of type will have a high probability of not making sense anymore - there is no way of knowing if the change proposed by Elastic will also apply to the field name that the user is now referring through its customization.
+
+Therefore, we should prefer to keep the user's version for the whole block.(See this example in the first element of the array in the first row of the table below).
+
+Examples: `related_integrations,` `required_fields`
+
+###### Proposed algorithm
+
+- Do element by element comparison:
+  - If element index exists in the three versions and the values:
+    - do not differ between all versions (base, current and target), **keep current** version. (A A A)
+    - do not differ between base and current, but updates target, **keep target** version. (A A B)
+    - differ between base and current, but doesn't update target, **keep current** version. (A B A)
+    - differ between base and current, updates target, **keep current** version. (A B B)
+    - differ between all version (base, current and target), **keep current** version. (A B C)
+  - If element index exists in base and current version, but not in target:
+    - If base differs from current: **keep current** version. (A B _)
+    - If base equals current: **remove element**.  (A A _)
+  - If element index exists in base and target version, but not in current:
+    - **Remove element**. (A _ B) and (A _ A)
+  - If element index exists in current and target version, but not in base:
```

Comment:
We shouldn't be comparing elements in the base, current and target arrays index-by-index. At least in most cases. In most cases, the order of elements doesn't matter: the order of tags in the tags array, the order of required fields, the order of related integrations, etc.

For example, these arrays of required fields should look completely equivalent to the diff algorithm:

```
base: [
  { ecs: true, name: 'host.name', type: 'keyword' },
  { ecs: true, name: 'host.ip', type: 'ip' },
  { ecs: false, name: 'custom_field', type: 'text' }
]
current: [
  { ecs: true, name: 'host.ip', type: 'ip' },
  { ecs: true, name: 'host.name', type: 'keyword' },
  { ecs: false, name: 'custom_field', type: 'text' }
]
target: [
  { ecs: true, name: 'host.name', type: 'keyword' },
  { ecs: false, name: 'custom_field', type: 'text' },
  { ecs: true, name: 'host.ip', type: 'ip' }
]
```

Which probably means the following:

- When order doesn't matter, we should be either a) using sets or maps instead of arrays under the hood when calculating diffs, or b) sorting arrays before comparing them.
- For some complex object values that don't have an explicit id, we should define an object's key for the diff algo. For example:
  - for required fields the key could be the name of the field
  - for related integrations the key could be the package name + the optional integration name
- For complex obects that already have a clear explicit id, we should be using this id as the object's key.
- The algorithm should be tracking changes to elements in the arrays by their keys.

The algorithm could have a similar structure to the one for arrays of scalar values (strings, numbers, etc):

- Given base is [element with key `A`, element with key `B`]
- Calculate changes to `current` compared against `base`:
  - Added: [element with key `C`]
  - Removed: [element with key `A`]
  - Modified: [element with key `B`]
- Changes to `target` compared against `base`:
  - Added: [element with key `D`]
  - Removed: [element with key `A`]
  - Modified: [element with key `B`]
- In combination:
  - Added: [element with key `C`, element with key `D`] (no conflicts)
  - Removed: [element with key `A`] (no conflicts)
  - Modified: [element with key `B`, element with key `B`] (conflict if these 2 are not exactly the same modification)

Reply:
Reworked whole section using this logic.

I have the same question for the removal of elements that I wrote in https://github.com/elastic/kibana/pull/171856#discussion_r1537371632